### ISSUE TITLE
Support self-referential generic implicit/PEP-613 type aliases

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#1738]
We currently handle self-referential generic implicit (and PEP 613) type aliases by [falling back to `Divergent`](https://github.com/astral-sh/ruff/blob/f4e4229683936a486f689aebb9d7d06b3985952d/crates/ty_python_semantic/resources/mdtest/implicit_type_aliases.md?plain=1#L1532-L1548). This prevents infinite recursion and false positive diagnostics, but we should properly support them. For example, this should be an error, but currently is not:
```py
from typing import TypeVar

T = TypeVar("T")
NestedDict = dict[str, "NestedDict[T]"]

n: NestedDict[int] = {"foo": b"wrong"}
```

An example case where this is relied on in typeshed is the annotation for the second argument to `isinstance`, which is a recursive tuple-of-tuples. Currently we don't catch the type error in `isinstance(None, (int, None))` (though we do catch it in `isinstance(None, None)`), due to not fully supporting the recursion.

(Note: we [already support this](https://play.ty.dev/22eb39cd-098f-4b84-80e9-70bc088740d5) for PEP 695 type aliases)

**Comment by mtshiba** [comment]
If we can support implicit recursive type aliases, we should also be able to automatically support PEP-613-style type aliases.
To support implicit recursive type aliases, they must be lazily evaluated if named.

**Comment by andyli-oai** [comment]
Declaration order of `JsonObject` and `JsonArray` changes whether TY catches invalid indexing (`ty 0.0.74 Linux`)

**Object alias first:**
```python
from typing import Mapping, Sequence, reveal_type

JsonObject = Mapping[str, "JsonValue"]
JsonArray = Sequence["JsonValue"]
JsonValue = None | int | JsonArray | JsonObject


def example(data: JsonObject) -> None:
    reveal_type(data["items"])    # Divergent
    data["items"][0]              # ❌ No errors reported.
```
**Array alias first:**
```python
from typing import Mapping, Sequence, reveal_type

JsonArray = Sequence["JsonValue"]
JsonObject = Mapping[str, "JsonValue"]
JsonValue = None | int | JsonArray | JsonObject

def example(data: JsonObject) -> None:
    reveal_type(data["items"])    # None | int | Sequence[Divergent] | Mapping[str, Divergent]
    data["items"][0]              # ✅ catches 3 errors: None, int are not subscriptable. Mapping requires a str key.
```
Expected: both versions should report the errors.

**Comment by carljm** [comment]
@mtshiba Is this something you are still actively working on?

**Comment by mtshiba** [comment]
> [@mtshiba](https://github.com/mtshiba) Is this something you are still actively working on?

Yes, but my current prototype, while able to correctly inspect the code presented here, is causing several stack overflows and hangs.
It opens up a path for recursive type inference that has previously been closed due to dynamic types, so it is likely hitting a potential recursion bug. I am currently investigating the cause.
